rfcs/2021-06-01-7694-improve-monitor-build-performance.md

```diff
@@ -0,0 +1,115 @@
+
+We should be running the Datadog agent on every runner possible, collecting both high-level system metrics as well as more fine-grained metrics, such as Docker metrics on a per-container basis.  These metrics will, I believe, prove invaluable to debugging issues with the overall CI pipeline and its performance over time.  While understanding the build performance will likely come down to more advanced tooling used directly by engineers, we can’t optimize things like “use all the cores available” or “we’re using too much memory during CI runs” unless we actually have some data, any data, from those runs.
+
+### Manually instrument CI to report build performance
+
+Once we were properly utilizing Datadog on our CI runners, we could then instrument CI builds
+directly to report the build time to Datadog, allowing us to track the build performance over time.
+
+This data would be sparse given how many builds are likely to be triggered on a given day, but would
+form the start of tracking build performance holistically.
+
+## Rationale
+
+Currently, as mentioned above, it can take nearly 40 minutes for a clean release build of Vector.
+Depending on the hardware used, this can be reduced to 25 - 30 minutes, but the time spent is
+untenable even at the 25 - 30 minute range.  This build time not only permeates the local
+development experience, but everything happening in CI, as well.  Any step of the lifecycle of a
+change that involves Vector is unnecessarily slowed down.  The build time of Vector is a massive
+force multiplier.
+
+If we changed nothing about the process, we could still successfully work on Vector, albeit with a
+slower rate of change.  Given the need to occasionally pivot to new features, or bug fixes, or
+whatever is the most important task at hand, rebuilding Vector is actually more common than one
+might expect, and thus every move we make -- whether it be working on a new feature, or fixing a bug
+-- will be slowed down by suboptimal build performance.
+
+## Prior Art
+
+Existing prior art matches the path being proposed here.  There is no one solution to build
+performance, as it is a function of many different settings working in concert.
+
+## Drawbacks
+
+There is a risk that the implementation of this proposal potentially results in occasional
+drawbacks, namely two: `sccache` poisoned cache, and unobservable performance regressions.
+
+As `sccache` caches compiled dependencies, there is the risk that a dependency could be compiled
+such that it is misconfigured, or if it was compiled with bad settings, it would then be reused even
+after the issue was fixed, leading to a confusing development experience.  Practically speaking, it
+is simple to clear the cache, but it would be a new potential line item for things to check when
+debugging an unclear/confusing compilation error, or functional bug.
```

Comment:
Is `sccache` aware of our operating systems? IIRC we have Mac OS on x66 and ARM soon, a handful of Linux x86 machines in the project. About the only thing I like about bazel is the ability for a cache to be declared "upstream", such that a dev machine can _pull_ from the upstream cache (refreshed by CI) but cannot _push_ into it. Makes poisoning rare.

Reply:
To my knowledge, it _should_ handle these differences, but their documentation is essentially non-existent on this point.  I'll dig deeper into this.

Reply:
We might want to consider explicitly avoiding the build cache when producing release builds, to better ensure reproducibility.

Reply:
Totally agreed re: reproducible builds.